### ISSUE TITLE
Built-in FS monitor daemon maintains a lock on worktree directories such that they cannot be moved

### ISSUE THREAD

**Issue by vermiculus** [git-for-windows/git#3408]
- [x] I was not able to find an [open](https://github.com/git-for-windows/git/issues?q=is%3Aopen) or [closed](https://github.com/git-for-windows/git/issues?q=is%3Aclosed) issue matching what I'm seeing

### Setup

 - Which version of Git for Windows are you using? Is it 32-bit or 64-bit?

```
$ git --version --build-options

git version 2.33.0.windows.1
cpu: x86_64
built from commit: fb59c1de9fd599a3eeb8459eb7cab731c6672ea3
sizeof-long: 4
sizeof-size_t: 8
shell-path: /bin/sh
feature: fsmonitor--daemon
```

 - Which version of Windows are you running? Vista, 7, 8, 10? Is it 32-bit or 64-bit?

```
$ cmd.exe /c ver

Microsoft Windows [Version 10.0.19042.1110]

```

 - What options did you set as part of the installation? Or did you choose the
   defaults?

```
# One of the following:
> type "C:\Program Files\Git\etc\install-options.txt"
> type "C:\Program Files (x86)\Git\etc\install-options.txt"
> type "%USERPROFILE%\AppData\Local\Programs\Git\etc\install-options.txt"
$ cat /etc/install-options.txt

Editor Option: CustomEditor
Custom Editor Path: C:\emacs\bin\emacsclient.exe
Default Branch Option: main
Path Option: Cmd
SSH Option: OpenSSH
Tortoise Option: false
CURL Option: WinSSL
CRLF Option: LFOnly
Bash Terminal Option: MinTTY
Git Pull Behavior Option: Merge
Use Credential Manager: Core
Performance Tweaks FSCache: Enabled
Enable Symlinks: Disabled
Enable Pseudo Console Support: Disabled
Enable FSMonitor: Enabled
```

 - Any other interesting things about your environment that might be related
   to the issue you're seeing?

Nothing comes to mind.

### Details

 - Which terminal/shell are you running Git from? e.g Bash/CMD/PowerShell/other

Bash

 - What commands did you run to trigger this issue? If you can provide a
   [Minimal, Complete, and Verifiable example](http://stackoverflow.com/help/mcve)
   this will help us understand the issue.

```
# ensure the built-in fs monitor is enabled, then
git clone foo.git && cd foo
git worktree add ../my-branhc my-branch
git worktree move ../my-branhc ../my-branch
```
 - What did you expect to occur after running these commands?

I expected the worktree to be moved without issue.

 - What actually happened instead?

```
$ git worktree move ../fooo/ ../foo
Rename from '/c/path/to/git/my-branhc' to '../my-branch' failed. Should I try again? (y/n)
```
We've determined that this is because the fsmonitor is maintaining a lock on the directory. If I `git fsmonitor--daemon stop` inside the worktree directory, the move succeeds.

 - If the problem was occurring with a specific repository, can you provide the
   URL to that repository to help us with testing?

N/A

----

Recommend stopping the fsmonitor and restarting it again after the move.

Let me know if this should be submitted upstream instead; I know Git4Win is using its own built-in fsmonitor here for ease-of-setup, but I'm not sure where to draw the lines between what can be fixed here vs within git.git.

Thanks!

**Comment by vermiculus** [comment]
Thanks Jeff :-)

Is there a mailing list thread I can subscribe myself to for updates on the upstream fix? (Assuming such a thread has even been started; I'm not sure on where you are in the development process for it.) Once the fix is in place upstream, I'll want to start the process on my end to get everyone upgraded and get this configuration in place. I can also help be an advocate for the change if that would be helpful :-)

Unless I misunderstand the nature of the fix you're planning -- if this thread will be notified once the fix is in place for GFW even before it lands on git.git, I can leave it at that.

**Comment by dscho** [comment]
> Is there a mailing list thread I can subscribe myself to for updates on the upstream fix?

Here is the PR, which gets updated when the patches get discussed on the Git mailing list or integrated into one of Git's integration branches: https://github.com/gitgitgadget/git/pull/923

> once the fix is in place for GFW even before it lands on git.git

We will be likely to update Git for Windows' [snapshots](https://wingit.blob.core.windows.net/files/index.html) (by virtue of integration newer patch series iterations into Git for Windows' `main` branch). Meaning: the `main` branch will be the thing to look out for: https://github.com/git-for-windows/git/commits/main

**Comment by jeffhostetler** [comment]
I have sent V3 of this feature to the mailing list in [1].  The code and conversation for this can be see in [2].

We shipped an experimental preview of my WIP V4 in git-for-windows v2.33 that included some mailing list feedback and feedback from an earlier preview that we shipped in git-for-windows v2.32.

I'm currently finalizing my V4 which will address all of the feedback received so far and will (hopefully) send that to the mailing list this week.  I can't say how long it will need to cook on the mailing list, but we will take that version of the code and update the experimental preview in the v2.34 version of git-for-windows.

[1] https://lore.kernel.org/git/pull.923.v3.git.1625150864.gitgitgadget@gmail.com/
[2] https://github.com/gitgitgadget/git/pull/923

Thanks for helping us beat on this new feature and let us know if you have further problems.
Jeff
